### PR TITLE
Made LinearOperator independent of spLinearOperator

### PR DESCRIPTION
**Pull request by rohanbabbar04**
For issue #328 
- Added `_AdjointLinearOperator`, `_TransposeLinearOperator` and `_ProductLinearOperator`
- Added methods `__matmul__`,`__rmatmul__` , `transpose` and `adjoint`
- Made all classes extend the `LinearOperator`

Still need to change/update the docs, if code seems fine...

### CODE REVIEW

**Comment by mrava87** [comment]
Thank you!

I'll try to review this in coming days and let you know :)

File: pylops/linearoperator.py
Comment (mrava87) [inline pylops/linearoperator.py:316]: I don't understand these changes.. in the past, we needed to check if an operator was a spicy operator and wrap it into a pylops linear operator... this is what `aslinearoperator` is mostly for... 

This:
```
if isinstance(y, LinearOperator):
    y = aslinearoperator(y)
```
doesn't really make any sense to me as if you look at what `aslinearoperator` does this is the code:

```
if isinstance(Op, LinearOperator):
   return Op
else:
   return LinearOperator(Op)
```
but since you already do the same check you are guaranteed already that the return will be Op

File: pylops/linearoperator.py
Comment (mrava87) [inline pylops/linearoperator.py:309]: There is no more need for this...

File: pylops/linearoperator.py
Comment (mrava87) [inline pylops/linearoperator.py:337]: Now that we implement all private methods, there is no more need for this...

This comment applies in many places :)

**Review by mrava87 — COMMENTED** [review]
Here are some comments. I have also made a commit where I propose changes according to my comments. 

@rohanbabbar04, feel free to go over and let me know what you think.

File: pylops/linearoperator.py
Comment (rohanbabbar04) [reply pylops/linearoperator.py]: Updated the docs, do take a look and give your opinion..

File: pylops/linearoperator.py
Comment (rohanbabbar04) [reply pylops/linearoperator.py:286]: Agree with your change as Op._matvec will be much faster

File: pylops/linearoperator.py
Comment (rohanbabbar04) [reply pylops/linearoperator.py:337]: Noted and agreed

File: pylops/linearoperator.py
Comment (rohanbabbar04) [reply pylops/linearoperator.py:95]: Noted finally it is `Op:Union[spLinearOperator, LinearOperator]`

**Comment by rohanbabbar04** [comment]
Thank You @mrava87 for reviewing my PR, I have updated the docs and agree with all your changes.
Do take a look and give your opinion

**Comment by mrava87** [comment]
> Thank You @mrava87 for reviewing my PR, I have updated the docs and agree with all your changes.
> Do take a look and give your opinion

Will do soon :)

**Comment by mrava87** [comment]
I just identified a problem with `describe` due to some of the internal linop methods being moved from scipy to native pylops ones...

For example: 
```
A = pylops.MatrixMult(np.ones((10, 5)))
A.name = "A"
C = pylops.MatrixMult(np.ones((10, 5)))

# Sum
D = A + C
describe(D)
```

now returns `A where: {'A': '_SumLinearOperator'}` instead of `A+M where: {'A': 'MatrixMult', 'M': 'MatrixMult'}`. 

Currently investigating why...

File: pylops/linearoperator.py
Comment (mrava87) [reply pylops/linearoperator.py:95]: No is `Optional[Union[spLinearOperator, LinearOperator]]`

**Comment by rohanbabbar04** [comment]
Is the bug fixed?

**Comment by rohanbabbar04** [comment]
Just tested the code 
Now the describe works fine
```
A + M
where: {'A': 'MatrixMult', 'M': 'MatrixMult'}
```

**Comment by rohanbabbar04** [comment]
> Just tested the code Now the describe works fine
> 
> ```
> A + M
> where: {'A': 'MatrixMult', 'M': 'MatrixMult'}
> ```

So I am removing the docstrings to make the `Codacy Static Code Analysis` successful
Removing this and making a commit

**Comment by mrava87** [comment]
Yes the code is fixed :)

**Comment by rohanbabbar04** [comment]
> Yes the code is fixed :)

Made the commit all checks are now successful...

**Review by mrava87 — APPROVED** [review]
I made some extra changes as you completely removed any mention to scipy LinearOperator, which instead is worth still mentioning (also because we can still mix our operators with theirs - and with their solvers!)

**Comment by mrava87** [comment]
@rohanbabbar04 I am satisfied with the current status of the PR.

Since we are making changes to a core method of PyLops, I want to get the feedback of another core developer before merging.. please be patient, we will let you know soon when this is ready to go :)

**Comment by cako** [comment]
I'll be trying this patch very soon! Sorry for the delays!

**Review by cako — APPROVED** [review]
This is great stuff. Looks good to me!!

**Comment by cako** [comment]
closes #328

**Pull request merged**
